### PR TITLE
Add trivial memtier kind policy

### PR DESCRIPTION
**Pull request by michalbiesek**
Recommend to review commit by commit

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/571)

### CODE REVIEW

**Review by PatKamin — COMMENTED** [review]
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/571)* status: all files reviewed, 1 unresolved discussion (waiting on @michalbiesek)

___
*[src/memkind_memtier.c, line 151 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtea3Z5kb_ZL9NEkUs:-MVtea3_2WVo7PDG53Bb:b-gxarm6) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/src/memkind_memtier.c#L151)):*
> ```C
>         (*kind)->builder->cfg[i].tier_ratio = builder->cfg[i].tier_ratio;
>     }
>     tier_id = 0;
> ```

I guess it's temporary?

File: src/memkind_memtier.c [hunk 1]
```diff
@@ -92,18 +95,22 @@ MEMKIND_EXPORT int memtier_builder_set_policy(struct memtier_builder *builder,
...
     return -1;
 }
 
-static inline memkind_t get_memtier_kind(const struct memtier_kind *tier_kind)
+static inline memkind_t get_memtier_kind(struct memtier_kind *tier_kind)
 {
-    // TODO now assign first memory kind - fix it with policy logic
-    return tier_kind->builder->cfg[0].tier->kind;
+    if (tier_kind->builder->policy == MEMTIER_POLICY_CIRCULAR) {
+        unsigned temp_id = (tier_id++) % tier_kind->builder->size;
```
Comment (PatKamin) [inline src/memkind_memtier.c:109]: Do the check: `size > 0`

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/571)* status: all files reviewed, 4 unresolved discussions (waiting on @michalbiesek)

___
*[src/memkind_memtier.c, line 203 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtxkik4_7KK7Eh_B--:-MVtxkik4_7KK7Eh_B-0:b1q38ku) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/src/memkind_memtier.c#L203)):*
> ```C
>         return memtier_tier_realloc(tier, ptr, size);
>     }
>     return memkind_realloc(get_memtier_kind(kind), ptr, size);
> ```

if ptr == null we could call malloc here
___
*[test/memkind_memtier_test.cpp, line 154 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtzLGc5DuAkFw7K5My:-MVtzLGc5DuAkFw7K5Mz:b-18xa12) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/test/memkind_memtier_test.cpp#L154)):*
> ```cpp
> TEST_F(MemkindMemtierBuilderTest, test_tier_builder_allocation_test_success)
> {
>     const size_t size = 512;
> ```

This test is valid only for "circular" policy. And if you use m_tier_kind only here, I think it would be best to move building of it here.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/571)* status: 0 of 9 files reviewed, 4 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[src/memkind_memtier.c, line 109 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVqhA2f5i65eL2TEGXf-r3-109:-MVuAAKVEjDXDKROzEtL:b9023sp) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/src/memkind_memtier.c#L109)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Do the check: `size > 0`
</blockquote></details>

Not needed to, see the construct kind limitation
___
*[src/memkind_memtier.c, line 151 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtea3Z5kb_ZL9NEkUs:-MVuAEbF3CA-2FzYpq7Z:b-jukc6d) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/src/memkind_memtier.c#L151)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

I guess it's temporary?
</blockquote></details>

Yes this is a dummy politic that will be later removed
___
*[src/memkind_memtier.c, line 203 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtxkik4_7KK7Eh_B--:-MVuAH_hEPUH-N_-PeT5:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/src/memkind_memtier.c#L203)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

if ptr == null we could call malloc here
</blockquote></details>

Done.
___
*[test/memkind_memtier_test.cpp, line 154 at r3](https://reviewable.io/reviews/memkind/memkind/571#-MVtzLGc5DuAkFw7K5My:-MVuBtvUC255W4CkELBT:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/1680671c1d55179ba8b8c58538898b3cd4fb8b53/test/memkind_memtier_test.cpp#L154)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

This test is valid only for "circular" policy. And if you use m_tier_kind only here, I think it would be best to move building of it here. 
</blockquote></details>

Done.

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> Reviewed 9 of 9 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, 5 of 5 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/571#-:-MVuJ4PiBcmLjgtt-r_D:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @PatKamin)

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/571#-:-MVutI0l0GkoogRrt2rI:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @michalbiesek)

**Pull request merged**